sockeye/data_io.py

```diff
@@ -388,90 +1027,171 @@ def get_parallel_bucket(buckets: List[Tuple[int, int]],
+        return ParallelDataSet(source, target, label)
+
+    def fill_up(self,
+                bucket_batch_sizes: List[BucketBatchSize],
+                fill_up: str,
+                seed: int = 42) -> 'ParallelDataSet':
+        """
+        Returns a new dataset with buckets filled up using the specified fill-up strategy.
+
+        :param bucket_batch_sizes: Bucket batch sizes.
+        :param fill_up: Fill-up strategy.
+        :return: New dataset with buckets filled up to the next multiple of batch size
+        """
+        source = list(self.source)
+        target = list(self.target)
+        label = list(self.label)
+
+        rs = np.random.RandomState(seed)
+
+        for bucket_idx in range(len(self)):
+            bucket = bucket_batch_sizes[bucket_idx].bucket
+            bucket_batch_size = bucket_batch_sizes[bucket_idx].batch_size
+            bucket_source = self.source[bucket_idx]
+            bucket_target = self.target[bucket_idx]
+            bucket_label = self.label[bucket_idx]
+            num_samples = bucket_source.shape[0]
+
+            if num_samples % bucket_batch_size != 0:
+                if fill_up == 'replicate':
+                    rest = bucket_batch_size - num_samples % bucket_batch_size
+                    logger.info("Replicating %d random samples from %d samples in bucket %s "
+                                "to size it to multiple of %d",
+                                rest, num_samples, bucket, bucket_batch_size)
+                    random_indices = mx.nd.array(rs.randint(num_samples, size=rest))
+                    source[bucket_idx] = mx.nd.concat(bucket_source, bucket_source.take(random_indices), dim=0)
+                    target[bucket_idx] = mx.nd.concat(bucket_target, bucket_target.take(random_indices), dim=0)
+                    label[bucket_idx] = mx.nd.concat(bucket_label, bucket_label.take(random_indices), dim=0)
+                else:
+                    raise NotImplementedError('Unknown fill-up strategy')
+
+        return ParallelDataSet(source, target, label)
+
+    def permute(self, permutations):
+        assert len(self) == len(permutations)
+        source = []
+        target = []
+        label = []
+        for buck_idx in range(len(self)):
+            num_samples = self.source[buck_idx].shape[0]
+            if num_samples:  # not empty bucket
+                permutation = permutations[buck_idx]
+                source.append(self.source[buck_idx].take(permutation))
+                target.append(self.target[buck_idx].take(permutation))
+                label.append(self.label[buck_idx].take(permutation))
+            else:
+                source.append(self.source[buck_idx])
+                target.append(self.target[buck_idx])
+                label.append(self.label[buck_idx])
+        return ParallelDataSet(source, target, label)
+
+
+def get_permutations(bucket_counts: List[int]) -> Tuple[List[mx.nd.array], List[mx.nd.array]]:
```

Comment:
This could use a docstring:
```python
"""
Returns the indices of a random permutation for each bucket and the corresponding inverse permutations that would restore the original order of the data if applied to the permuted data.
"""
```

Reply:
done